File: src/main/java/net/igneo/icv/enchantmentActions/enchantManagers/weapon/MeathookManager.java
package net.igneo.icv.enchantmentActions.enchantManagers.weapon;

import net.igneo.icv.ICV;
import net.igneo.icv.enchantment.EnchantType;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.phys.Vec3;

import java.util.HashMap;

public class MeathookManager extends WeaponEnchantManager {
    public MeathookManager(Player player) {
        super(EnchantType.WEAPON, player, new ResourceLocation(ICV.MOD_ID, "dual_handed_slash_cross"));
    }
    
    private final HashMap<Entity, Integer> meat = new HashMap<>();
    
    @Override
    public void onAttack(Entity entity) {
        super.onAttack(entity);
        meat.put(entity, 200);
    }
    
    @Override
    public void tick() {
        super.tick();
        for (Entity entity : meat.keySet()) {
            meat.replace(entity, meat.get(entity) - 1);
            if (meat.get(entity) <= 0) {
                meat.remove(entity);
            }
        }
    }
    
    @Override
    public void activate() {
        super.activate();
        double x = 0;
        double y = 0;
        double z = 0;
        for (Entity entity : meat.keySet()) {
            x += entity.getX();
            y += entity.getY();
            z += entity.getZ();
        }
        x = x / meat.keySet().size();
        y = y / meat.keySet().size();
        z = z / meat.keySet().size();
        Vec3 average = new Vec3(x, y, z);
        
        for (Entity entity : meat.keySet()) {
            if (entity.distanceToSqr(average) < 70) {
                Vec3 pushVec = entity.position().subtract(average).normalize().reverse();
                entity.addDeltaMovement((entity.onGround() ? new Vec3(pushVec.x, 0.5, pushVec.z) : pushVec));
            }
        }
        
        meat.clear();
    }
}

File: src/main/java/net/igneo/icv/enchantmentActions/enchantManagers/weapon/MeathookManager.kt
package net.igneo.icv.enchantmentActions.enchantManagers.weapon

import net.igneo.icv.ICV
import net.igneo.icv.enchantment.EnchantType
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.player.Player
import net.minecraft.world.phys.Vec3

class MeathookManager(player: Player?) :
    WeaponEnchantManager(
        EnchantType.WEAPON,
        player!!,
        ResourceLocation(ICV.MOD_ID, "dual_handed_slash_cross")
    ) {
    private val meat = HashMap<Entity, Int>()

    override fun onAttack(entity: Entity?) {
        super.onAttack(entity)
        meat[entity!!] = 200
    }

    override fun tick() {
        super.tick()
        for (entity in meat.keys) {
            meat.replace(entity, meat[entity]!! - 1)
            if (meat[entity]!! <= 0) {
                meat.remove(entity)
            }
        }
    }

    override fun activate() {
        super.activate()
        var x = 0.0
        var y = 0.0
        var z = 0.0
        for (entity in meat.keys) {
            x += entity.x
            y += entity.y
            z += entity.z
        }
        x /= meat.keys.size
        y /= meat.keys.size
        z /= meat.keys.size
        val average = Vec3(x, y, z)

        for (entity in meat.keys) {
            if (entity.distanceToSqr(average) < 70) {
                val pushVec = entity.position().subtract(average).normalize().reverse()
                entity.addDeltaMovement((if (entity.onGround()) Vec3(pushVec.x, 0.5, pushVec.z) else pushVec))
            }
        }

        meat.clear()
    }
}
